src/main/java/uk/nhs/hee/trainee/details/api/BasicDetailsResource.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright 2020 Crown Copyright (Health Education England)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+ * associated documentation files (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge, publish, distribute,
+ * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.trainee.details.api;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PatchMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import uk.nhs.hee.trainee.details.dto.PersonalDetailsDto;
+import uk.nhs.hee.trainee.details.mapper.PersonalDetailsMapper;
+import uk.nhs.hee.trainee.details.model.PersonalDetails;
+import uk.nhs.hee.trainee.details.service.PersonalDetailsService;
+
+@Slf4j
+@RestController
+@RequestMapping("/api/basic-details")
+public class BasicDetailsResource {
+
+  private PersonalDetailsService service;
+  private PersonalDetailsMapper mapper;
+
+  public BasicDetailsResource(PersonalDetailsService service, PersonalDetailsMapper mapper) {
+    this.service = service;
+    this.mapper = mapper;
+  }
+
+  /**
+   * Update the person details for the trainee, creates the parent profile if it does not already
+   * exist.
+   *
+   * @param tisId The ID of the trainee to update.
+   * @param dto   The person details to update with.
+   * @return The updated or created PersonalDetails.
+   */
+  @PatchMapping("/{tisId}")
+  public ResponseEntity<PersonalDetailsDto> updateBasicDetails(
+      @PathVariable(name = "tisId") String tisId, @RequestBody PersonalDetailsDto dto) {
+    log.trace("Update basic details of trainee with TIS ID {}", tisId);
+    PersonalDetails entity = mapper.toEntity(dto);
+    entity = service.createProfileOrUpdateBasicDetailsByTisId(tisId, entity);
```

Comment:
I'm on mobile... Should this endpoint be able to create basic details when they don't exist when it's a PATCH request?

Reply:
It should, its the lesser evil versus doubling the number of requests.
When the Person table is being synced we need to take two actions:

1.  Create a profile if it doesn't already exist
2.  Update the basic details in an existing profile

PATCH is used because the PersonalDetailsDto is shared across the different endpoints, which only returns the data available in the single TIS table. We could use a DTO per update type I guess.

Reply:
I guess it depends how pure you want these restful services to be. I'm fine with bending the rules every now and then

Reply:
Pepe found a line in the spec where it said Patch **may** create a resource, so we went with it.

Reply:
Ah, did not know that. Thought it was supposed to fail if the resource didnt exist